python/pyspark/sql/datasource.py

```diff
@@ -403,3 +410,36 @@ def register(
         assert sc._jvm is not None
         ds = sc._jvm.org.apache.spark.sql.execution.python.UserDefinedPythonDataSource(wrapped)
         self.sparkSession._jsparkSession.dataSource().registerPython(name, ds)
+
+
+class CaseInsensitiveDict(UserDict):
+    """
+    A case-insensitive map of string keys to values.
+
+    This is used by Python data source options to ensure consistent case insensitivity.
+
+     .. versionadded: 4.0.0
```

Comment:
qq why should we have this as an API? Thought users won't create this instance manually.

Reply:
Yea you are right. Users shouldn't directly use this (unless they want to unit test the data source).